### ISSUE TITLE
Hooks fail to run with "Bad system call" due to SystemCallFilter in zrepl.service.

### ISSUE THREAD

**Issue by juskglass** [zrepl/zrepl#740]
Example output from logs:
```
zrepl[4462]: [snap_backup][hook][Uv38$Ypc+$Wgkr$Wgkr]: hook output stderr="/etc/zrepl/hooks/nextcloud.sh: line 6: 28438 Bad system call         $DRYRUN sudo -u www-data /usr/bin/php /var/www/nextcloud/occ maintenance:mode --on" snap="zrepl_20231002_233004_000" command="/etc/zrepl/hooks/nextcloud.sh" fs="tank/ds1/alpha/nextcloud"
```

https://github.com/zrepl/zrepl/blob/c0b52b92d5c8caf52e03d4d332cc06e13a11d3b9/dist/systemd/zrepl.service#L30-L34

Initially, I disabled the filter with `systemctl edit zrepl.service`, adding the following to `override.conf`:
```
[Service]
SystemCallFilter=
``` 

In theory, I could try to identify which system calls the hook commands rely on and add them to the filter. I did not attempt this. Instead, I reverted the change above and used a pair of oneshot services, one each for the pre- and post-snapshot hooks, and updated my hook script to start those services. 

This seemed like a better solution: if desired, one could add appropriate values for `SystemCallFilter` (and other sandboxing options) to each service, as opposed to trying to add everything to `zrepl.service`. The downside is that every hook script needs a pair of services to go with it which will only ever be used by that script.

Is there a better way to handle this? What's the recommended configuration for hook scripts that fail due to the `SystemCallFilter`?

**Comment by problame** [comment]
Good that you understood what the problem was and found a workaround. cc #735 

>  This seemed like a better solution: if desired, one could add appropriate values for SystemCallFilter (and other sandboxing options) to each service, as opposed to trying to add everything to zrepl.service. The downside is that every hook script needs a pair of services to go with it which will only ever be used by that script.

Interesting, let me paraphrase to ensure I understood correctly: the default .service file's permissions allow running `systemctl start whatever-oneshot.service`. So, you defined your hooks as oneshot systemd services and the zrepl config just launches a tiny script that does `systemctl start pre-oneshot.service`.

Correctly understood?

> Is there a better way to handle this? What's the recommended configuration for hook scripts that fail due to the SystemCallFilter?

Honestly I think the `override.conf` is fine.
Your solution preserves the SystemCallFilter, which is nice in theory.
But, ultimately, if the hook can invoke `systemctl`, it probably also can invoke `systemd-run` (?).
An attacker who controls the zrepl process could use systemd-run to run an arbitrary command, without the SystemCallFilter limitations.

I haven contemplated many times whether the repeated trouble with the `SystemCallFilter` directive is worth it.

**Comment by juskglass** [comment]
> Interesting, let me paraphrase to ensure I understood correctly: the default .service file's permissions allow running `systemctl start whatever-oneshot.service`. So, you defined your hooks as oneshot systemd services and the zrepl config just launches a tiny script that does `systemctl start pre-oneshot.service`.
> 
> Correctly understood?

Correct. `pre_snapshot()` and `post_snapshot()` in the hook script each contain a single line: `systemctl start [pre|post]-snapshot.service`.

> > Is there a better way to handle this? What's the recommended configuration for hook scripts that fail due to the SystemCallFilter?
> 
> Honestly I think the `override.conf` is fine. Your solution preserves the SystemCallFilter, which is nice in theory. But, ultimately, if the hook can invoke `systemctl`, it probably also can invoke `systemd-run` (?). An attacker who controls the zrepl process could use systemd-run to run an arbitrary command, without the SystemCallFilter limitations.

Yes, I was able to replace the oneshot services with equivalent `systemd-run` commands, eliminating the need for the unit files. Presumably there's some way to prevent calling `systemd-run`, otherwise `SystemCallFilter` seems to be of limited use. I don't know it though, and I didn't see anything obvious is a quick look through the documentation.
